pkg/kubelet/kubelet_node_status.go

```diff
@@ -238,13 +238,23 @@ func (kl *Kubelet) initialNode() (*v1.Node, error) {
 		Effect: v1.TaintEffectNoSchedule,
 	}
 
-	// If TaintNodesByCondition enabled, taint node with TaintNodeUnschedulable when initializing
+	notReadyTaint := v1.Taint{
+		Key:    schedulerapi.TaintNodeNotReady,
+		Effect: v1.TaintEffectNoSchedule,
+	}
+
+	// If TaintNodesByCondition enabled, taint node with TaintNodeUnschedulable and TaintNodeNotReady when initializing
 	// node to avoid race condition; refer to #63897 for more detail.
 	if utilfeature.DefaultFeatureGate.Enabled(features.TaintNodesByCondition) {
 		if node.Spec.Unschedulable &&
 			!taintutil.TaintExists(nodeTaints, &unschedulableTaint) {
 			nodeTaints = append(nodeTaints, unschedulableTaint)
 		}
+		// Add not-ready taint while initializing the node to avoid waiting for node controller to apply the taint.
+		// This can be removed by node controller once the node is ready.
```

Comment:
> This is done at initialization. What happens when the node goes notReady for some reason while kubelet is up ?

The Node will still have not-ready taint.

Adding not-ready taint at Node creation is in order to mark Node as pristine, which is never processed by node-lifecycle-controller, and scheduler should nenver treate it as schedulable. 

As after node-lifecycle-controller processed, the Node taints is credible (even it also had some race issue as you mentioned), while before node-lifecycle-controller processed, the Node lack some condition -> taints process, scheduler must treate it as unschedulable.

Reply:
> What happens when the node goes notReady for some reason while kubelet is up ?

This case could still happen. The problem at a fundamental level is kubelet is updating node conditions within kubelet code and taints reflecting those conditions are updated within node lifecycle controller(not in kubelet). During this time the pods would get scheduled onto that node but eventually node lifecycle controller would add a `NoExecute` taint to the node which will cause those pods to evict. So, this is kind of inline with kubernetes way of doing things but not super elegant from end-user perspective.

Reply:
sgtm

Reply:
> During this time the pods would get scheduled onto that node but eventually node lifecycle controller would add a NoExecute taint to the node which will cause those pods to evict.

The default timeout of `NoExecute` is 300 seconds; that means the service maybe unavailable until node ready or got evicted after 300 seconds. I don-t think online service can tolerate such long period :)

Reply:
I think it be more if we also consider the node-monitor-grace-period. But indeed this might be a bit much.
I was considering also another approach where taints  would be timestamped and having a recovery mechanism based on that. But this requires some changes at the API level.

Reply:
> Either adding taints to node.Status or having a recovery mechanism need API level changes.

How about tolerate conditions? 

I'd suggest someone can help to drive a brainstoming to enhance that, e.g. the cases we meet, candidate options, compatibility. If no volunteer, I'd like to contribute :)

Reply:
That's a good idea @k82cn. I am fine with starting a discussion around this. I will send a google document with list of options discussed so far and people can expand on it.